net.bobbo.player-controller/player_scripts/crouch.gd

```diff
@@ -33,4 +41,35 @@ func player_physics_process(delta: float):
 	var scale_difference = new_scale - previous_scale
 	
 	player.velocity.y += scale_difference * 0.5
+	
+	# Make it so that the crouch can only go as low as regularly crouching and only high enough such
+	# that the player doesn't get stuck if an object is detected above them
 	player.height = new_scale
+
+func _should_force_crouch():
+	# Get the 3D physics space from a Node3D like the player_pivot
+	var space_state = player.model_pivot.get_world_3d().direct_space_state
+	
+	var raycast_position_offset = 0
+	if(_should_be_crouched):
+		raycast_position_offset = _default_player_height / 4
+	
+	# Calculate origin and ending of raycast
+	var origin = player.position + player.up_direction * raycast_position_offset
+	var end = origin + player.up_direction * _default_player_height / 2
```

Comment:
So while testing I found that when player.height changes it also changes the vertical position of the player and since we want the raycast to originate from the same spot so we can consistently detect an obstruction above the player and that spot is based off the vertical position of the player it needs to be there in the case of the player crouching to account for the vertical position changing. If you were to remove it you would see that when uncrouching it won't keep the player down since the raycast would no longer be able to detect the surface above the player so then it causes the player to uncrouch and then crouch again over and over until there isn't an object above the player.

Reply:
> So while testing I found that when player.height changes it also changes the vertical position of the player...

This is correct, because the player's origin is the center of the collider - so when the collider gets shorter, the whole player gets physically closer to the ground.

> ...and since we want the raycast to originate from the same spot so we can consistently detect an obstruction above the player and that spot is based off the vertical position of the player it needs to be there in the case of the player crouching to account for the vertical position changing....

This is not necessarily true though. IMO we shouldn't cast from where the player _would_ be if they were standing simply because they're not standing. In theory if the player's crouch scale was lower than 50%, then this would even cause a bug - because the origin of the cast would be higher than the player's collider, potentially popping through a ceiling collider.

I got my hands dirty and created a visualization of this raycasting. Here is what it currently looks like:

https://github.com/Iseeicy/netengine5/assets/1341063/c78e4dc8-d5ba-48b7-94c7-020ec6a5e8af




> ...If you were to remove it you would see that when uncrouching it won't keep the player down since the raycast would no longer be able to detect the surface above the player so then it causes the player to uncrouch and then crouch again over and over until there isn't an object above the player.

This intrigued me because of my above disagreement, so I re-used the visualization to see what was happening:

https://github.com/Iseeicy/netengine5/assets/1341063/c0c9f8c0-27ce-457c-8382-ec8804a5a71b

It looks like the only problem is that the raycast doesn't go far enough...! You want this cast to end where the top of the player's collider would be, IF they were standing - but the origin should always be the center of the player. I think this will fix a few bugs with this.

Reply:
I think also instead of subtracting the offset you want to add it, so that the raycast goes slightly above the player collider.